internal/repository/file.go

```diff
@@ -0,0 +1,255 @@
+//
+// Since 2.0.0
+func (r *FileRepository) Writer(u fyne.URI) (fyne.URIWriteCloser, error) {
+	return openFile(uri, true)
+}
+
+// CanWrite implements repository.WriteableRepository.CanWrite
+//
+// Since 2.0.0
+func (r *FileRepository) CanWrite(u fyne.URI) (bool, error) {
+	f, err := os.OpenFile(u.Path(), os.WRONLY, 0666)
+	defer f.Close()
+
+	if os.IsPermission(err) {
+		return false, nil
+	}
+
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// Delete implements repository.WriteableRepository.Delete
+//
+// Since 2.0.0
+func (r *FileRepository) Delete(u fyne.URI) error {
+	return os.Remove(u.Path())
+}
+
+// Parent implements repository.HierarchicalRepository.Parent
+//
+// Since 2.0.0
+func (r *FileRepository) Parent(u fyne.URI) (fyne.URI, error) {
+	s := u.String()
+
+	// trim trailing slash
+	if s[len(s)-1] == '/' {
+		s = s[0 : len(s)-1]
+	}
+
+	// trim the scheme
+	s = s[len(u.Scheme())+3:]
+
+	// Completely empty URI with just a scheme
+	if len(s) == 0 {
+		return nil, repository.URIRootError
+	}
+
+	parent := ""
+	if u.Scheme() == "file" {
+		// use the system native path resolution
+		parent = filepath.Dir(s)
+		if parent[len(parent)-1] != filepath.Separator {
+			parent += "/"
+		}
+
+		// only root is it's own parent
+		if filepath.Clean(parent) == filepath.Clean(s) {
+			return nil, repository.URIRootError
+		}
+
+	} else {
+		var err error
+		parent, err = parentGeneric(s)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return ParseURI(u.Scheme() + "://" + parent)
+}
+
+// Child implements repository.HierarchicalRepository.Child
+//
+// Since 2.0.0
+func (r *FileRepository) Child(u fyne.URI, component string) (fyne.URI, error) {
+	// TODO: make sure that this works on Windows - might cause trouble
+	// if the path sep isn't normalized out on ingest.
+	return repository.GenericChild(u, component, storage.ParseURI)
+}
+
+// List implements repository.HierarchicalRepository.List()
+//
+// Since 2.0.0
+func (r *FileRepository) List(u fyne.URI) ([]fyne.URI, error) {
+	if u.Scheme() != "file" {
+		return nil, fmt.Errorf("unsupported URL protocol")
+	}
+
+	path := u.Path()
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	urilist := []fyne.URI{}
+
+	for _, f := range files {
+		uri, err := storage.ParseURI("file://" + filepath.Join(path, f.Name()))
+		if err != nil {
+			return nil, err
+		}
+		urilist = append(urilist, uri)
+	}
+
+	return urilist, nil
+}
+
+// CanList implements repository.HierarchicalRepository.CanList()
+//
+// Since 2.0.0
+func (r *FileRepository) CanList(u fyne.URI) (bool, error) {
+	info, err := os.Stat(u.Path())
+
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+
+	if err != nil {
+		return false, err
+	}
+
+	if !info.IsDir() {
+		return false, nil
+	}
+
+	// We know it is a directory, but we don't know if we can read it, so
+	// we'll just try to do so and see if we get a permissions error.
+	_, err := ioutil.ReadDir(path)
```

Comment:
Consider replacing this call to `ioutil.ReadDir` with something along the lines of the following, which will not only give you a more specific error, but also runs much faster (approximately half the ns per operation on my machine.)

```go
	// Open directory at path and read one entry from it
	f, err := os.Open(path)
	if err == nil {
		_, err = f.Readdir(1)
		f.Close()
	}
	
	// ... check err
```

Reply:
31da787f8dc500ac4b148d38e278e7f8342cd7e8